Sprint 7

dmirain opened the pull request:
Я не удержался и переписал логику на async/await и сделал так, что параллельно гружу все фотки, что бы пользователь не ждал между переключениями. 
Сделал рефакторинг и написал тесты почти на всё.
Что бы тесты не ходили в сеть настроил окружение tests и для него запускаю приложение с замоконным клиентом.

Commit:
Переписал работу с сетью на async/await. Гружу картинки параллельно.

Commit:
Поправил сильные ссылки на слабые.

Commit:
Более корректная обработка ошибок

Commit:
Правильное имя параметра

Commit:
Отрефакторил, выделив Presenter. Переделал инициализацию контроллера, создаю зависимоста в init.

Commit:
Упростил init контроллера

Commit:
Unit тесты на получение данных, на генерацию загадок и на расчёт статистики

Commit:
ArrayTest

Commit:
В окружении тестов использую мок данные

Commit:
UI тесты

Commit:
Тест на основную модель

amardanov commented on MovieQuiz/MovieQuizModel/MovieQuizModelImpl.swift:
```diff
@@ -1,16 +1,17 @@
 final class MovieQuizModelImpl: MovieQuizModel {
-    private weak var delegate: MovieQuizModelDelegat?
+
+    weak var delegate: MovieQuizModelDelegat?
```
в названии `MovieQuizModelDelegat` пропущено "e" в конце

amardanov commented on MovieQuizTests/KPGatewayTest.swift:
```diff
@@ -0,0 +1,75 @@
+import XCTest
+@testable import MovieQuiz
+
+struct MockHttp: NetworkClient {
```
Лучше всегда моки выносить в отдельные файлы, чтобы 
1) Не захламлять файл с тестами
2) Не забывать как называется мок и не искать его, чтобы переиспользовать при необходимости

amardanov commented on MovieQuizTests/MovieQuizModelTest.swift:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+
+    func testExample() throws {
```
не очень понятное название теста

amardanov commented on MovieQuizTests/MovieQuizModelTest.swift:
```diff
@@ -0,0 +1,120 @@
+import XCTest
+@testable import MovieQuiz
+
+struct MockStatisticService: StatisticService {
```
Тоже самое про моки в отдельных файлах

amardanov commented on MovieQuizTests/RiddleFactoryTest.swift:
```diff
@@ -0,0 +1,34 @@
+import XCTest
+@testable import MovieQuiz
+
+struct MockMovieHub: MovieHubGateway {
```
Тоже самое про моки в отдельных файлах

amardanov commented on MovieQuizTests/StatisticServiceTest.swift:
```diff
@@ -0,0 +1,65 @@
+import XCTest
+@testable import MovieQuiz
+
+struct MockStorageFirstGame: StatisticStorage {
```
Тоже самое про моки в отдельных файлах

Commit:
Исправил замечания ревью
- Поправил название MovieQuizModelDelegat -> MovieQuizModelDelegate
- Полностью перенёс работу с MovieQuizModel в MovieQuizPresenter
- Блокирую кнопки и проверяю это в тестах
- Вынес моки в отдельные файлы
- Исправил название теста testExample -> testGameLogic

Commit:
Немного красоты

dmirain commented:
@amardanov У меня тесты проходят стабильно. Попросил других прогнать для надёжности, тоже проходят. Хочется понять в каких условиях они падают у вас. Отправил ещё раз на проверку, поправив все замечания. Надеюсь с тестами получится разобраться.

Commit:
Убрал лишний параметр в инициализации MovieQuizPresenterImpl
